internal/versions/hash-generator/generate.go

```diff
@@ -19,52 +27,111 @@ import (
 	"log"
 	"net/http"
 	"os"
+	"slices"
 	"strings"
 
+	"github.com/regclient/regclient"
+	"github.com/regclient/regclient/types/ref"
+	"github.com/vincent-petithory/dataurl"
 	"golang.org/x/tools/go/ast/astutil"
+	"k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	"k8s.io/kubernetes/cmd/kubeadm/app/images"
 )
 
-func mustGetHash(url string) string {
-	// remove quotes around url
-	url = url[1 : len(url)-1]
+const (
+	defaultRegistry = "registry.k8s.io"
+	etcdComponent   = "etcd"
+	defaultFilePath = "./versions.go"
+)
+
+var supportedComponents = []string{"kube-apiserver", "kube-controller-manager", "kube-scheduler", "etcd"}
+
+func quote(s string) string {
+	return fmt.Sprintf(`"%s"`, s)
+}
+
+func unquote(s string) string {
+	return strings.TrimPrefix(strings.TrimSuffix(s, `"`), `"`)
+}
 
-	// Get the data
-	req, err := http.NewRequestWithContext(context.Background(), http.MethodGet, url, nil)
+// pinKubernetesImage takes a component and a version and returns the corresponding container image pinned by hash.
+//
+// The version string is a Kubernetes version tag, which is used to derive the tags of the component images.
+// The image hash is obtained directly from the default registry, registry.k8s.io.
+func pinKubernetesImage(comp, ver string) (string, error) {
+	if !slices.Contains(supportedComponents, comp) {
+		return "", fmt.Errorf("k8s component %q not supported: valid components: %#v", comp, supportedComponents)
+	}
+	ref := ref.Ref{Scheme: "reg", Registry: defaultRegistry, Repository: comp, Tag: ver}
+	if comp == etcdComponent {
+		cfg := &kubeadm.ClusterConfiguration{
+			KubernetesVersion: ver,
+			ImageRepository:   defaultRegistry,
+		}
+
+		img := images.GetEtcdImage(cfg)
+		_, tag, _ := strings.Cut(img, ":")
+		ref.Tag = tag
+	}
+	log.Printf("Getting hash for image %#v", ref)
+
+	rc := regclient.New()
+	m, err := rc.ManifestGet(context.Background(), ref)
 	if err != nil {
-		panic(err)
+		return "", fmt.Errorf("could not obtain image manifest: %w", err)
 	}
-	resp, err := http.DefaultClient.Do(req)
+
+	return fmt.Sprintf("%s/%s:%s@%s", ref.Registry, ref.Repository, ref.Tag, m.GetDescriptor().Digest.String()), nil
+}
+
+func generateKubeadmPatch(comp, ver string) (string, error) {
+	img, err := pinKubernetesImage(comp, ver)
 	if err != nil {
-		panic(err)
+		return "", err
+	}
+	content, err := json.Marshal([]map[string]string{{
+		"op":    "replace",
+		"path":  "/spec/containers/0/image",
+		"value": img,
+	}})
+	if err != nil {
+		return "", err
+	}
+	return dataurl.New(content, "application/json").String(), nil
+}
+
+// hashURLContent downloads a binary blob from the given URL and calculates its SHA256 hash.
+//
+// URLs passed to this function are expected to have upstream signatures with a .sha256 suffix. This upstream signatue
+// will be verified, too.
+func hashURLContent(url string) (string, error) {
+	resp, err := http.Get(url)
```

Comment:
The linter wants to use a version with ctx

Reply:
I fixed that by ignoring the linter, because I believe it does not apply in our situtation here. Please let me know whether you agree.